tgui/packages/tgui/interfaces/MusicInstrument.tsx

```diff
@@ -0,0 +1,183 @@
+import { useBackend, useLocalState } from '../backend';
+import { Button, Flex, LabeledList, Box, Knob, Input } from '../components';
+import { Window } from '../layouts';
+
+type MusicInstrumentData = {
+  name: string;
+  notes: string[];
+};
+
+export const MusicInstrument = (props, context) => {
+  const { act, data } = useBackend<MusicInstrumentData>(context);
+  const { name, notes } = data;
+
+  const [noteKeysOrder, setNoteKeysOrder] = useLocalState(
+    context,
+    'keyboardBindingsOrder',
+    'zsxdcvgbhnjmq2w3er5t6y7ui9o0p+'.split('')
+  );
+
+  const [activeKeys, setActiveKeys] = useLocalState(context, 'keyboardActivekeys', new Array(notes.length));
+  const [keyOffset, setKeyOffset] = useLocalState(context, 'keyOffset', 0);
+  const [keybindToggle, setKeybindToggle] = useLocalState(context, 'keybindToggle', false);
+  const [volume, setVolume] = useLocalState(context, 'keyboardVolume', 50);
+  const [transpose, setTranspose] = useLocalState(context, 'keyboardTranspose', 0);
+
+  const toggleKeybind = () => {
+    if (keybindToggle) {
+      act('play_keyboard_off');
+      setKeybindToggle(false);
+    } else {
+      act('play_keyboard_on');
+      setKeybindToggle(true);
+    }
+  };
+
+  const keyIndexWithinRange = (index: number) => index + transpose >= 0 && index + transpose < notes.length;
+
+  const playNote = (index: number) => {
+    if (keyIndexWithinRange(index) && !activeKeys[index]) {
+      act('play_note', { note: index + transpose + 1, volume: volume });
+      const newKeys = activeKeys;
+      newKeys[index] = true;
+      setActiveKeys(newKeys);
```

Comment:
This mutates `activeKeys` directly (then does it via the state setter).

Consider:
```suggestion
      const newKeys = [...activeKeys];
      newKeys[index] = true;
      setActiveKeys(newKeys);
```